File: libsignal-service/src/test/java/org/whispersystems/signalservice/internal/push/GroupMismatchedDevicesTest.java
package org.whispersystems.signalservice.internal.push;

import org.junit.Test;
import org.whispersystems.signalservice.internal.util.JsonUtil;

import java.io.IOException;

import static org.junit.Assert.assertEquals;

public class GroupMismatchedDevicesTest {

  @Test
  public void testSimpleParse() throws IOException {
    GroupMismatchedDevices[] parsed = JsonUtil.fromJson("[\n" +
        "  {\n" +
        "    \"uuid\": \"12345678-1234-1234-1234-123456789012\",\n" +
        "    \"devices\": {\n" +
        "      \"missingDevices\": [1, 2],\n" +
        "      \"extraDevices\": [3]\n" +
        "    }\n" +
        "  },\n" +
        "  {\n" +
        "    \"uuid\": \"22345678-1234-1234-1234-123456789012\",\n" +
        "    \"devices\": {\n" +
        "      \"missingDevices\": [],\n" +
        "      \"extraDevices\": [2]\n" +
        "    }\n" +
        "  }\n" +
        "]", GroupMismatchedDevices[].class);

    assertEquals(2,  parsed.length);
    assertEquals("12345678-1234-1234-1234-123456789012", parsed[0].getUuid());
    assertEquals(1, (int) parsed[0].getDevices().getMissingDevices().get(0));
    assertEquals(2, (int) parsed[0].getDevices().getMissingDevices().get(1));
    assertEquals(3, (int) parsed[0].getDevices().getExtraDevices().get(0));

    assertEquals("22345678-1234-1234-1234-123456789012", parsed[1].getUuid());
    assertEquals(2, (int) parsed[1].getDevices().getExtraDevices().get(0));
  }
}

File: lib/libsignal-service/src/test/java/org/whispersystems/signalservice/internal/push/GroupMismatchedDevicesTest.kt
package org.whispersystems.signalservice.internal.push

import assertk.assertThat
import assertk.assertions.containsExactly
import assertk.assertions.hasSize
import assertk.assertions.isEqualTo
import org.junit.Test
import org.whispersystems.signalservice.internal.util.JsonUtil

class GroupMismatchedDevicesTest {
  @Test
  fun testSimpleParse() {
    val json = """
      [
        {
          "uuid": "12345678-1234-1234-1234-123456789012",
          "devices": {
            "missingDevices": [1, 2],
            "extraDevices": [3]
          }
        },
        {
          "uuid": "22345678-1234-1234-1234-123456789012",
          "devices": {
            "missingDevices": [],
            "extraDevices": [2]
          }
        }
      ]
    """.trimIndent()
    val parsed = JsonUtil.fromJson(json, Array<GroupMismatchedDevices>::class.java)

    assertThat(parsed).hasSize(2)
    val (first, second) = parsed

    assertThat(first.uuid).isEqualTo("12345678-1234-1234-1234-123456789012")
    assertThat(first.devices.missingDevices).containsExactly(1, 2)
    assertThat(first.devices.extraDevices).containsExactly(3)

    assertThat(second.uuid).isEqualTo("22345678-1234-1234-1234-123456789012")
    assertThat(second.devices.extraDevices).containsExactly(2)
  }
}
